In [ ]:
from PIL import Image

# Load the image
img_path = '/content/drive/MyDrive/Capstone/Deakin/furseal/masked_dataset/A/a_1.jpg'
img = Image.open(img_path)

# Print the size (width, height)
print("Image size (width x height):", img.size)


Image size (width x height): (5472, 3648)


In [ ]:
from PIL import Image

# Load the image
img_path = '/content/drive/MyDrive/Capstone/Deakin/furseal/masked_dataset/B/b_2.jpg'
img = Image.open(img_path)

# Print the size (width, height)
print("Image size (width x height):", img.size)


Image size (width x height): (5472, 3648)


very high-resolution image — it’s about 20 megapixels (5472 × 3648 = 19,961,856 pixels


In [ ]:
from PIL import Image
import numpy as np


img = Image.open('/content/drive/MyDrive/Capstone/Deakin/furseal/masked_dataset/A/a_1.jpg')
img = img.resize((8, 8))  # Resize to 8x8
img = img.convert('L')    # Convert to grayscale
img_array = np.array(img).astype(np.float32) / 255.0  # Normalize pixel values
feature_vector = img_array.flatten()

print("Feature vector shape:", feature_vector.shape)


Feature vector shape: (64,)


In [ ]:
from PIL import Image
import numpy as np

def preprocess_image(img_path, size=(4, 4)):
    img = Image.open(img_path).convert("L")  # Grayscale
    img = img.resize(size)
    img_array = np.array(img).astype(np.float32) / 255.0  # Normalize
    return img_array.flatten()  # Shape: (16,)

# Example
img_path = '/content/drive/MyDrive/Capstone/Deakin/furseal/masked_dataset/A/a_1.jpg'
feature_vector = preprocess_image(img_path)
print("Feature vector length:", len(feature_vector))


Feature vector length: 16


In [ ]:
import os
from PIL import Image
from pathlib import Path

# Set original and new dataset paths
base_path = "/content/drive/MyDrive/Capstone/Deakin/furseal"
original_dataset_path = os.path.join(base_path, "masked_dataset")
reduced_dataset_path = os.path.join(base_path, "reduced_masked_dataset_5x5")

# Define the target size
resize_size = (5, 5)

# Create new directory
os.makedirs(reduced_dataset_path, exist_ok=True)

# Process folders A to H
folders = [chr(c) for c in range(ord('A'), ord('H')+1)]

for folder in folders:
    original_folder = os.path.join(original_dataset_path, folder)
    reduced_folder = os.path.join(reduced_dataset_path, folder)
    os.makedirs(reduced_folder, exist_ok=True)

    # Process each image in the folder
    for filename in os.listdir(original_folder):
        if filename.lower().endswith((".jpg", ".jpeg", ".png")):
            try:
                img_path = os.path.join(original_folder, filename)
                img = Image.open(img_path).convert("L")  # Grayscale
                img = img.resize(resize_size)

                # Save the resized image
                save_path = os.path.join(reduced_folder, filename)
                img.save(save_path)

                print(f"Processed {folder}/{filename}")
            except Exception as e:
                print(f"Failed to process {folder}/{filename}: {e}")


Processed A/a_1.jpg
Processed A/a_2.jpg
Processed A/a_3.jpg
Processed A/a_4.jpg
Processed A/a_5.jpg
Processed A/a_6.jpg
Processed A/a_7.jpg
Processed A/a_aug_1.jpg
Processed A/a_aug_2.jpg
Processed A/a_aug_3.jpg
Processed B/b_1.jpg
Processed B/b_2.jpg
Processed B/b_3.jpg
Processed B/b_4.jpg
Processed B/b_5.jpg
Processed B/b_6.jpg
Processed B/b_7.jpg
Processed B/b_aug_1.jpg
Processed B/b_aug_2.jpg
Processed B/b_aug_3.jpg
Processed C/c_1.jpg
Processed C/c_2.jpg
Processed C/c_3.jpg
Processed C/c_4.jpg
Processed C/c_5.jpg
Processed C/c_6.jpg
Processed C/c_7.jpg
Processed C/c_aug_1.jpg
Processed C/c_aug_2.jpg
Processed C/c_aug_3.jpg
Processed D/d_1.jpg
Processed D/d_2.jpg
Processed D/d_3.jpg
Processed D/d_4.jpg
Processed D/d_5.jpg
Processed D/d_6.jpg
Processed D/d_7.jpg
Processed D/d_8.jpg
Processed D/d_9.jpg
Processed D/d_10.jpg
Processed E/e_1.jpg
Processed E/e_2.jpg
Processed E/e_3.jpg
Processed E/e_4.jpg
Processed E/e_5.jpg
Processed E/e_6.jpg
Processed E/e_7.jpg
Processed E/e_8.jpg
Pro

/content/drive/MyDrive/Capstone/Deakin/furseal/reduced_masked_dataset_5x5


In [ ]:
from PIL import Image
import numpy as np

def preprocess_image(img_path, size=(4, 4)):
    img = Image.open(img_path).convert("L")  # Grayscale
    img = img.resize(size)
    img_array = np.array(img).astype(np.float32) / 255.0  # Normalize
    return img_array.flatten()  # Shape: (16,)

# Example
img_path = '/content/drive/MyDrive/Capstone/Deakin/furseal/reduced_masked_dataset_5x5/A/a_1.jpg'
feature_vector = preprocess_image(img_path)
print("Feature vector length:", len(feature_vector))

Feature vector length: 16


In [ ]:
import os
from PIL import Image
from pathlib import Path

# Set original and new dataset paths
base_path = "/content/drive/MyDrive/Capstone/Deakin/furseal"
original_dataset_path = os.path.join(base_path, "masked_dataset")
reduced_dataset_path = os.path.join(base_path, "reduced_masked_dataset_128x128")

# Target resize dimensions
resize_size = (128, 128)

# Create base output folder
os.makedirs(reduced_dataset_path, exist_ok=True)

# Folder names: A to H
folders = [chr(c) for c in range(ord('A'), ord('H')+1)]

for folder in folders:
    src_folder = os.path.join(original_dataset_path, folder)
    dst_folder = os.path.join(reduced_dataset_path, folder)
    os.makedirs(dst_folder, exist_ok=True)

    # Process each image file
    for filename in os.listdir(src_folder):
        if filename.lower().endswith((".jpg", ".jpeg", ".png")):
            try:
                img_path = os.path.join(src_folder, filename)
                img = Image.open(img_path).convert("L")  # Convert to grayscale
                img = img.resize(resize_size)  # Resize to 128x128

                # Save processed image
                save_path = os.path.join(dst_folder, filename)
                img.save(save_path)

                print(f"Saved: {folder}/{filename}")
            except Exception as e:
                print(f"Error processing {folder}/{filename}: {e}")


Saved: A/a_1.jpg
Saved: A/a_2.jpg
Saved: A/a_3.jpg
Saved: A/a_4.jpg
Saved: A/a_5.jpg
Saved: A/a_6.jpg
Saved: A/a_7.jpg
Saved: A/a_aug_1.jpg
Saved: A/a_aug_2.jpg
Saved: A/a_aug_3.jpg
Saved: B/b_1.jpg
Saved: B/b_2.jpg
Saved: B/b_3.jpg
Saved: B/b_4.jpg
Saved: B/b_5.jpg
Saved: B/b_6.jpg
Saved: B/b_7.jpg
Saved: B/b_aug_1.jpg
Saved: B/b_aug_2.jpg
Saved: B/b_aug_3.jpg
Saved: C/c_1.jpg
Saved: C/c_2.jpg
Saved: C/c_3.jpg
Saved: C/c_4.jpg
Saved: C/c_5.jpg
Saved: C/c_6.jpg
Saved: C/c_7.jpg
Saved: C/c_aug_1.jpg
Saved: C/c_aug_2.jpg
Saved: C/c_aug_3.jpg
Saved: D/d_1.jpg
Saved: D/d_2.jpg
Saved: D/d_3.jpg
Saved: D/d_4.jpg
Saved: D/d_5.jpg
Saved: D/d_6.jpg
Saved: D/d_7.jpg
Saved: D/d_8.jpg
Saved: D/d_9.jpg
Saved: D/d_10.jpg
Saved: E/e_1.jpg
Saved: E/e_2.jpg
Saved: E/e_3.jpg
Saved: E/e_4.jpg
Saved: E/e_5.jpg
Saved: E/e_6.jpg
Saved: E/e_7.jpg
Saved: E/e_8.jpg
Saved: E/e_9.jpg
Saved: E/e_10.jpg
Saved: F/f_1.jpg
Saved: F/f_2.jpg
Saved: F/f_3.jpg
Saved: F/f_4.jpg
Saved: F/f_5.jpg
Saved: F/f_6.jpg
Saved: F/f

/content/drive/MyDrive/Capstone/Deakin/furseal/reduced_masked_dataset_128x128

In [ ]:
from PIL import Image

# Load the image
img_path = '/content/drive/MyDrive/Capstone/Deakin/furseal/reduced_masked_dataset_128x128/B/b_1.jpg'
img = Image.open(img_path)

# Print the size (width, height)
print("Image size (width x height):", img.size)


Image size (width x height): (128, 128)


autoencoder

In [ ]:
import os
import numpy as np
from PIL import Image

# Image folder
data_path = "/content/drive/MyDrive/Capstone/Deakin/furseal/reduced_masked_dataset_128x128/"

image_size = 128  # assuming grayscale 128x128
X = []
y = []

label_map = {folder: idx for idx, folder in enumerate(sorted(os.listdir(data_path)))}

for folder, label in label_map.items():
    folder_path = os.path.join(data_path, folder)
    for file in os.listdir(folder_path):
        img = Image.open(os.path.join(folder_path, file)).convert("L").resize((image_size, image_size))
        arr = np.array(img).flatten()
        arr = arr / np.linalg.norm(arr)  # normalize for amplitude encoding
        X.append(arr)
        y.append(label)

X = np.array(X)
y = np.array(y)
print(f"Loaded {X.shape[0]} samples of dimension {X.shape[1]}")


Loaded 80 samples of dimension 16384


In [ ]:
from PIL import Image
import os
import numpy as np

data_path = "/content/drive/MyDrive/Capstone/Deakin/furseal/reduced_masked_dataset_128x128/"

shapes = []

for folder in sorted(os.listdir(data_path)):
    folder_path = os.path.join(data_path, folder)
    for file in os.listdir(folder_path):
        img = Image.open(os.path.join(folder_path, file)).convert("L")
        shapes.append(np.array(img).shape)
        break  # Just check one per folder

print("Sample image shapes from each folder:")
print(shapes)


Sample image shapes from each folder:
[(128, 128), (128, 128), (128, 128), (128, 128), (128, 128), (128, 128), (128, 128), (128, 128)]


In [ ]:
import os
from PIL import Image
import numpy as np

data_path = "/content/drive/MyDrive/Capstone/Deakin/furseal/reduced_masked_dataset_128x128/"

X = []
y = []
label_map = {}

for label_idx, folder in enumerate(sorted(os.listdir(data_path))):
    label_map[label_idx] = folder  # 0: 'A', 1: 'B', ...
    folder_path = os.path.join(data_path, folder)

    for file in os.listdir(folder_path):
        img_path = os.path.join(folder_path, file)

        # Load and preprocess
        img = Image.open(img_path).convert("L").resize((16, 8))  # Resize to 128 pixels
        vec = np.array(img).flatten().astype(np.float32)
        vec = vec / np.linalg.norm(vec)  # Normalize for amplitude encoding

        X.append(vec)
        y.append(label_idx)

X = np.array(X)
y = np.array(y)

print(f"Shape of X: {X.shape}, Shape of y: {y.shape}")
print(f"Label mapping: {label_map}")


Shape of X: (80, 128), Shape of y: (80,)
Label mapping: {0: 'A', 1: 'B', 2: 'C', 3: 'D', 4: 'E', 5: 'F', 6: 'G', 7: 'H'}


In [ ]:
# Install compatible versions of Qiskit and machine learning modules
!pip install qiskit==0.44 qiskit-machine-learning==0.6.0 qiskit-aer


  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.4/133.4 kB 9.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
INFO: pip is looking at multiple versions of qiskit-aer to determine which version is compatible with other requirements. This could take a while.
INFO: pip is still looking at multiple versions of qiskit-aer to determine which version is compatible with other requirements. This could take a while.
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 148.7/148.7 kB 14.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 13.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
import os
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from qiskit import QuantumCircuit
from qiskit.circuit.library import StatePreparation
from qiskit.primitives import Sampler
from qiskit_machine_learning.kernels import FidelityQuantumKernel
from qiskit_machine_learning.kernels.algorithms import Fidelity
from qiskit_machine_learning.algorithms import PegasosQSVC


ImportError: cannot import name 'Sampler' from 'qiskit.primitives' (/usr/local/lib/python3.11/dist-packages/qiskit/primitives/__init__.py)

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

from qiskit import QuantumCircuit
from qiskit.circuit.library import ZZFeatureMap
from qiskit.utils import algorithm_globals
from qiskit_machine_learning.kernels import FidelityQuantumKernel
from qiskit_machine_learning.algorithms import PegasosQSVC


In [ ]:
import os

In [ ]:
from PIL import Image

# Constants
data_path = "/content/drive/MyDrive/Capstone/Deakin/furseal/reduced_masked_dataset_128x128/"
image_size = 128  # 128x128
max_dim = 2**7    # 128 features (7 qubits)

X, y = [], []

# Create label map A=0, B=1, ..., H=7
label_map = {folder: idx for idx, folder in enumerate(sorted(os.listdir(data_path)))}

for folder, label in label_map.items():
    folder_path = os.path.join(data_path, folder)
    for file in os.listdir(folder_path):
        img = Image.open(os.path.join(folder_path, file)).convert("L").resize((image_size, image_size))
        vec = np.array(img).flatten().astype(np.float32)
        vec = vec / np.linalg.norm(vec)  # amplitude encoding requires norm-1
        vec = vec[:max_dim]              # trim to 128 dims for 7 qubits
        X.append(vec)
        y.append(label)

X = np.array(X)
y = np.array(y)
print(f"Shape of encoded data: {X.shape}, Labels: {np.unique(y)}")


Shape of encoded data: (80, 128), Labels: [0 1 2 3 4 5 6 7]


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.25, random_state=42)


In [ ]:
n_qubits = 7  # Because 2^7 = 128

def amplitude_feature_map(x):
    qc = QuantumCircuit(n_qubits)
    qc.append(StatePreparation(x), qc.qubits)
    return qc


In [ ]:
from qiskit.primitives import Sampler


In [ ]:
print(dir(qmla))

NameError: name 'qmla' is not defined

In [ ]:
from qiskit_machine_learning.kernels.algorithms import Fidelity


ImportError: cannot import name 'Fidelity' from 'qiskit_machine_learning.kernels.algorithms' (/usr/local/lib/python3.11/dist-packages/qiskit_machine_learning/kernels/algorithms/__init__.py)

from qiskit.primitives import Sampler


In [ ]:
# Use Qiskit primitives
sampler = Sampler()
fidelity = Fidelity(sampler=sampler)

# Define fidelity kernel with amplitude encoding feature map
kernel = FidelityQuantumKernel(feature_map=amplitude_feature_map, fidelity=fidelity)

# Pegasos QSVC (Quantum SVM)
pegasos = PegasosQSVC(quantum_kernel=kernel, C=1.0)

# Train model
pegasos.fit(X_train, y_train)


NameError: name 'Fidelity' is not defined

In [ ]:
# Predict and score
y_pred = pegasos.predict(X_test)
acc = accuracy_score(y_test, y_pred)

print(f"PegasosQSVC accuracy with amplitude encoding: {acc:.4f}")


In [ ]:
!pip install -q --upgrade pip
!pip uninstall -y qiskit qiskit-machine-learning qiskit-aer qiskit-terra
!pip install qiskit==0.45.1 qiskit-machine-learning==0.6.1 qiskit-aer


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 32.4 MB/s eta 0:00:00
Found existing installation: qiskit 0.44.0
Uninstalling qiskit-0.44.0:
  Successfully uninstalled qiskit-0.44.0
Found existing installation: qiskit-machine-learning 0.6.0
Uninstalling qiskit-machine-learning-0.6.0:
  Successfully uninstalled qiskit-machine-learning-0.6.0
Found existing installation: qiskit-aer 0.13.1
Uninstalling qiskit-aer-0.13.1:
  Successfully uninstalled qiskit-aer-0.13.1
Found existing installation: qiskit-terra 0.25.0
Uninstalling qiskit-terra-0.25.0:
  Successfully uninstalled qiskit-terra-0.25.0
  Using cached qiskit_aer-0.17.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (8.2 kB)
INFO: pip is looking at multiple versions of qiskit-aer to determine which version is compatible with other requirements. This could take a while.
  Using cached qiskit_aer-0.16.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (8.2 kB)
  Using cached qiskit_aer-0.16

In [ ]:
# 🧠 Required Libraries
import os
import numpy as np
from PIL import Image
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

from qiskit import QuantumCircuit
from qiskit.circuit.library import StatePreparation
from qiskit.primitives import Sampler

from qiskit_machine_learning.kernels import FidelityQuantumKernel
from qiskit_machine_learning.kernels.algorithms import Fidelity
from qiskit_machine_learning.algorithms import PegasosQSVC


ImportError: cannot import name 'Fidelity' from 'qiskit_machine_learning.kernels.algorithms' (/usr/local/lib/python3.11/dist-packages/qiskit_machine_learning/kernels/algorithms/__init__.py)

In [ ]:
# 🔁 Clean reinstall of Qiskit packages
!pip uninstall -y qiskit qiskit-machine-learning qiskit-aer qiskit-terra
!pip install qiskit==0.45.1 qiskit-machine-learning==0.6.1 qiskit-aer

# 🔁 Restart runtime automatically
import os
os.kill(os.getpid(), 9)


Found existing installation: qiskit 0.45.1
Uninstalling qiskit-0.45.1:
  Successfully uninstalled qiskit-0.45.1
Found existing installation: qiskit-machine-learning 0.6.1
Uninstalling qiskit-machine-learning-0.6.1:
  Successfully uninstalled qiskit-machine-learning-0.6.1
Found existing installation: qiskit-aer 0.14.1
Uninstalling qiskit-aer-0.14.1:
  Successfully uninstalled qiskit-aer-0.14.1
Found existing installation: qiskit-terra 0.45.1
Uninstalling qiskit-terra-0.45.1:
  Successfully uninstalled qiskit-terra-0.45.1
  Using cached qiskit-0.45.1-py3-none-any.whl.metadata (12 kB)
  Using cached qiskit_machine_learning-0.6.1-py3-none-any.whl.metadata (12 kB)
  Using cached qiskit_aer-0.17.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (8.2 kB)
  Using cached qiskit_terra-0.45.1-cp38-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)
INFO: pip is looking at multiple versions of qiskit-aer to determine which version is compatible with other requi

In [ ]:
!pip show qiskit
!pip show qiskit-machine-learning


Name: qiskit
Version: 0.45.1
Summary: Software for developing quantum computing programs
Home-page: https://qiskit.org/
Author: Qiskit Development Team
Author-email: hello@qiskit.org
License: Apache 2.0
Location: /usr/local/lib/python3.11/dist-packages
Requires: qiskit-terra
Required-by: qiskit-aer
Name: qiskit-machine-learning
Version: 0.6.1
Summary: Qiskit Machine Learning: A library of quantum computing machine learning experiments
Home-page: https://github.com/Qiskit/qiskit-machine-learning
Author: Qiskit Machine Learning Development Team
Author-email: hello@qiskit.org
License: Apache-2.0
Location: /usr/local/lib/python3.11/dist-packages
Requires: dill, fastdtw, numpy, psutil, qiskit-terra, scikit-learn, scipy, setuptools
Required-by: 


In [ ]:
from qiskit.primitives import Sampler
from qiskit_machine_learning.kernels.algorithms import Fidelity

fidelity = Fidelity(sampler=Sampler())
print("✅ Fidelity is available and working.")


ImportError: cannot import name 'Fidelity' from 'qiskit_machine_learning.kernels.algorithms' (/usr/local/lib/python3.11/dist-packages/qiskit_machine_learning/kernels/algorithms/__init__.py)

In [ ]:

from qiskit_machine_learning.kernels.algorithms import Fidelity


ImportError: cannot import name 'Fidelity' from 'qiskit_machine_learning.kernels.algorithms' (/usr/local/lib/python3.11/dist-packages/qiskit_machine_learning/kernels/algorithms/__init__.py)

In [ ]:
from qiskit.primitives import Sampler
from qiskit_machine_learning.kernels.fidelity import Fidelity

sampler = Sampler()
fidelity = Fidelity(sampler=sampler)
print("✅ Fidelity is working.")


ModuleNotFoundError: No module named 'qiskit_machine_learning.kernels.fidelity'